File: app/core/roi.py
from typing import List, Optional, Sequence, Tuple

import cv2
import numpy as np


class ROI:
    """A polygonal region of interest defined by N corners in image coordinates.

    Corners connect in click order; the polygon is finalized when ``close()`` is called.
    Constructing with ``corners`` yields an already-closed ROI (the shape tools and tests
    build a complete polygon in one shot this way).
    """

    MIN_CORNERS = 3  # a usable polygon needs at least a triangle

    def __init__(self, corners: Optional[Sequence[Tuple[float, float]]] = None):
        self.corners: List[Tuple[float, float]] = [
            (float(x), float(y)) for x, y in (corners or [])
        ]
        self.closed: bool = bool(self.corners)

    @property
    def is_complete(self) -> bool:
        # Closed *and* at least a triangle: guards against a degenerate 1–2 corner polygon
        # (e.g. ROI([p]) ) being treated as a usable, fillable region.
        return self.closed and len(self.corners) >= self.MIN_CORNERS

    def add_corner(self, x: float, y: float) -> None:
        if not self.closed:
            self.corners.append((float(x), float(y)))

    def close(self) -> None:
        self.closed = True

    def reset(self) -> None:
        self.corners = []
        self.closed = False

    def _contour(self) -> np.ndarray:
        return np.array(self.corners, dtype=np.float32).reshape(-1, 1, 2)

    def polygon_int(self) -> np.ndarray:
        return np.array(self.corners, dtype=np.int32)

    def mask(self, height: int, width: int) -> np.ndarray:
        """Binary (0/255) uint8 mask of the filled polygon at image resolution."""
        m = np.zeros((height, width), dtype=np.uint8)
        if self.is_complete:
            cv2.fillPoly(m, [self.polygon_int()], 255)
        return m

    def contains(self, x: float, y: float) -> bool:
        if not self.is_complete:
            return False
        return cv2.pointPolygonTest(self._contour(), (float(x), float(y)), False) >= 0

    def contains_many(self, pts) -> np.ndarray:
        """Boolean mask over rows of ``pts`` (an ``(N, 2)`` array of ``(x, y)``) that lie inside a
        *complete* ROI. All-False for an incomplete ROI. Vectorized sibling of :meth:`contains`."""
        pts = np.asarray(pts, dtype=np.float32).reshape(-1, 2)
        if not self.is_complete:
            return np.zeros(len(pts), dtype=bool)
        contour = self._contour()
        return np.array(
            [cv2.pointPolygonTest(contour, (float(x), float(y)), False) >= 0 for x, y in pts],
            dtype=bool,
        )

File: rust/crates/core/src/roi.rs
//! Polygonal region of interest (corners in image coordinates).
//! Ports `app/core/roi.py`.

use opencv::core::{Mat, Point2f, Point2i, Scalar, Vector, CV_8UC1};
use opencv::imgproc;

/// A polygonal ROI defined by N corners in image coordinates. Constructing with
/// `corners` yields an already-closed polygon (how the shape tools/tests build one).
#[derive(Clone, Debug, Default)]
pub struct Roi {
    pub corners: Vec<(f64, f64)>,
    pub closed: bool,
}

impl Roi {
    pub const MIN_CORNERS: usize = 3;

    /// Already-closed ROI from a full corner list.
    pub fn new(corners: Vec<(f64, f64)>) -> Self {
        let closed = !corners.is_empty();
        Self { corners, closed }
    }

    /// Closed *and* at least a triangle (guards degenerate 1–2 corner polygons).
    pub fn is_complete(&self) -> bool {
        self.closed && self.corners.len() >= Self::MIN_CORNERS
    }

    pub fn add_corner(&mut self, x: f64, y: f64) {
        if !self.closed {
            self.corners.push((x, y));
        }
    }

    pub fn close(&mut self) {
        self.closed = true;
    }

    pub fn reset(&mut self) {
        self.corners.clear();
        self.closed = false;
    }

    fn contour(&self) -> Vector<Point2f> {
        self.corners
            .iter()
            .map(|&(x, y)| Point2f::new(x as f32, y as f32))
            .collect()
    }

    fn polygon_int(&self) -> Vector<Point2i> {
        self.corners
            .iter()
            .map(|&(x, y)| Point2i::new(x as i32, y as i32))
            .collect()
    }

    /// Binary (0/255) uint8 mask of the filled polygon at image resolution.
    pub fn mask(&self, height: i32, width: i32) -> opencv::Result<Mat> {
        let mut m = Mat::new_rows_cols_with_default(height, width, CV_8UC1, Scalar::all(0.0))?;
        if self.is_complete() {
            let mut polys = Vector::<Vector<Point2i>>::new();
            polys.push(self.polygon_int());
            imgproc::fill_poly_def(&mut m, &polys, Scalar::all(255.0))?;
        }
        Ok(m)
    }

    pub fn contains(&self, x: f64, y: f64) -> opencv::Result<bool> {
        if !self.is_complete() {
            return Ok(false);
        }
        let d = imgproc::point_polygon_test(&self.contour(), Point2f::new(x as f32, y as f32), false)?;
        Ok(d >= 0.0)
    }

    /// Boolean mask over `pts` that lie inside a *complete* ROI (all-false otherwise).
    pub fn contains_many(&self, pts: &[(f64, f64)]) -> opencv::Result<Vec<bool>> {
        if !self.is_complete() {
            return Ok(vec![false; pts.len()]);
        }
        let contour = self.contour();
        let mut out = Vec::with_capacity(pts.len());
        for &(x, y) in pts {
            let d = imgproc::point_polygon_test(&contour, Point2f::new(x as f32, y as f32), false)?;
            out.push(d >= 0.0);
        }
        Ok(out)
    }
}
